pkg/etcd/transport/server.go

```diff
@@ -0,0 +1,193 @@
+	respCh := make(chan *Response, len(agents))
+
+	for _, agent := range agents {
+		go func(agentName string) {
+			resp, err := t.SendRequest(agentName, msg)
+			if err != nil {
+				log.Error().Err(err).Msg("failed to send request to agent")
+				respCh <- &Response{
+					Client: agentName,
+					Error:  err,
+				}
+			} else {
+				respCh <- resp
+			}
+		}(agent)
+	}
+
+	resps := make([]Result[*RespValue], 0, len(agents))
+	for i := 0; i < len(agents); i++ {
+		resp := <-respCh
+		if resp.Error != nil {
+			return nil, resp.Error
+		}
+		var result RespValue
+		if err := proto.Unmarshal(resp.Data, Resp(&result)); err != nil {
+			return nil, status.Error(codes.InvalidArgument, err.Error())
+		}
+		resps = append(resps, Result[*RespValue]{
+			Client:  resp.Client,
+			Success: &result,
+		})
+	}
+
+	return resps, nil
+}
+
+// SendRequest allows consumers of the etcd transport to send single request to agents.
+func SendRequest[RespValue any, Resp ExactMessage[RespValue]](t *EtcdTransportServer, client string, msg proto.Message) (*RespValue, error) {
+	resp, err := t.SendRequest(client, msg)
+	if err != nil {
+		return nil, err
+	}
+
+	var result RespValue
+	if err := proto.Unmarshal(resp.Data, Resp(&result)); err != nil {
+		return nil, err
+	}
+
+	return &result, nil
+}
+
+// SendRequest sends a request to etcd, supposed to be consumed by an agent.
+func (t *EtcdTransportServer) SendRequest(client string, msg proto.Message) (*Response, error) {
+	anyreq, err := anypb.New(msg)
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+
+	rawReq, err := proto.Marshal(anyreq)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal request: %w", err)
+	}
+
+	req := Request{
+		ID:     uuid.NewString(),
+		Client: client,
+		Data:   rawReq,
+	}
+
+	path := path.Join(RPCBasePath, RPCRequestPath, req.Client, req.ID)
+
+	lease, err := t.etcdClient.Grant(context.Background(), 30)
+	if err != nil {
+		return nil, fmt.Errorf("failed to grant lease: %w", err)
+	}
+
+	_, err = t.etcdClient.Put(context.Background(), path, string(rawReq), clientv3.WithLease(lease.ID))
+	if err != nil {
+		return nil, fmt.Errorf("failed to send request to etcd: %w", err)
+	}
+
+	return t.waitForResponse(req)
+}
+
+func (t *EtcdTransportServer) waitForResponse(req Request) (*Response, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
+	defer cancel()
+
+	responePath := path.Join(RPCBasePath, RPCResponsePath, req.Client, req.ID)
+
+	resp, err := t.etcdClient.Get(ctx, responePath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get response from etcd: %w", err)
+	}
+
+	if len(resp.Kvs) > 0 {
+		return &Response{
+			Client: req.Client,
+			Data:   resp.Kvs[0].Value,
+		}, nil
+	}
+
+	watchCh := t.etcdClient.Watch(ctx, responePath)
+	for {
+		select {
+		case watchResp, ok := <-watchCh:
+			if !ok {
+				return nil, fmt.Errorf("watch channel closed")
+			}
+			if watchResp.Err() != nil {
```

Comment:
Should either do:
1. Just a watch
2. Periodic fetch (no watch)


Don't do fetch followed by watch

Reply:
removed the fetch